hyp3lib/get_orb.py

```diff
@@ -1,197 +1,147 @@
+        qc_data = response.json()
+
+        orbit_url = None
+        if qc_data["results"]:
+            orbit_url = qc_data["results"][0]["remote_url"]
+        return orbit_url
+
+    elif provider.upper() == 'ASF':
+        orbit_url = _get_asf_orbit_url(orbit_type.lower(), platform, time_stamps[0].replace('T', ''))
+        return orbit_url
+
+    raise OrbitDownloadError(f'Unknown orbit file provider {provider}')
+
+
+def _download_and_verify_orbit(url: str, directory: str = ''):
+    orbit_file = download_file(url, directory=directory, chunk_size=5242880)
     try:
-        stateVecFile=downloadSentinelOrbitFileProvider(granule,'ASF',directory)
-        provider = "ASF"
-        print("Found state vector file at ASF")
-    except:
-        print("Unable to find statevector at ASF; trying ESA")
+        verify_opod(orbit_file)
+    except ValueError:
+        logging.warning(f'Downloaded an invalid orbit file {orbit_file}')
+        return None
+
+    return orbit_file
+
+
+def downloadSentinelOrbitFile(granule: str, directory: str = '', providers=('ESA', 'ASF')):
+    """Download a Sentinel-1 Orbit file
+
+    Args:
+        granule: Granule name to find an orbit file for
+        directory: Directory to save the orbit files into
+        providers: Iterable of providers to attempt to download the orbit file from, in order of preference
+
+    Returns: Tuple of:
+        orbit_file: The downloaded orbit file
+        provider: The provider used to download the orbit file from
+
+    """
+    for provider in providers:
+        try:
+            url = get_orbit_url(granule, 'AUX_POEORB', provider=provider)
+            if url is not None:
+                orbit_file = _download_and_verify_orbit(url, directory=directory)
+                if orbit_file:
+                    return orbit_file, provider
+        except (requests.RequestException, OrbitDownloadError) as e:
+            logging.exception('Error encountered fetching orbit file; looking for another', exc_info=e)
+            continue
+
+    for provider in providers:
         try:
-            stateVecFile = downloadSentinelOrbitFileProvider(granule,'ESA',directory) 
-            if stateVecFile:
-                provider = "ESA"
-                print("Found state vector file at ESA")
-        except:
-            print("Unable to find requested state vector file")
-            provider = "NA"
-            return None,provider
-    verify_opod(stateVecFile)
-    return stateVecFile, provider
+            url = get_orbit_url(granule, 'AUX_RESORB', provider=provider)
+            if url is not None:
+                orbit_file = _download_and_verify_orbit(url, directory=directory)
+                if orbit_file:
+                    return orbit_file, provider
+        except (requests.RequestException, OrbitDownloadError) as e:
+            logging.exception('Error encountered fetching orbit file; looking for another', exc_info=e)
+            continue
```

Comment:
could be cleaner with a helper function:
```
for orbit_type in ('AUX_POEORB', 'AUX_RESORB'):
    orbit_file, provider = download_orbit_file_by_type(granule, orbit_type, providers=providers)
    if orbit_file:
        return orbit_file, providers
```

Reply:
I tired the helper function but I didn't like it -- pulled too much logic out of `downloadSentinelOrbitFile`. But, it def. looks better with orbit types as an outer for loop.